design/01-proc-macro.md

```diff
@@ -0,0 +1,299 @@
+#     serde_json = "1"
+
+[package]
+name = "ignore-plugin"
+version = "0.1.0"
+
+[lib]
+crate-type = "dylib"
+
+[dependencies]
+cairo-lang-macro-attributes = "1.0.0-alpha.7"
+cairo-lang-macro-interface = "1.0.0-alpha.7"
+cairo-lang-macro-interface-stable = "1.0.0-alpha.7"
+```
+
+### Procedural macro packages can be used as regular dependencies
+See [the guide-level explanation](#Procedural-macro-packages-can-be-used-as-dependencies) for more details.
+
+### Scarb will build and load procedural macros on user machines
+
+- Source distribution takes burden of building procedural macros from their authors.
+    - But it requires users to have Rust toolchain installed on their machines.
+        - Scarb itself will not contain any Rust source code compilation capabilities.
+        - Scarb will require users to have Cargo available, if it spots a procedural macro that has not been yet compiled, i.e. there is not a shared object/DLL binary in Scarb's cache.
+        - Thanks to this, writing code not using dynamic procedural macros will not require having `rustc` installed.
+- The procedural macros will be compiled with stable ABI layout of structs passing the FFI border. This should guarantee Rust ABI compatibility, regardless of Cargo toolchain version available on user machine. The `dylib` crate type will be safe, and thus this should prevent runtime crashes.
+- Running Rust compiler, and storing `target` directory will be completely private thing to Scarb.
+    - Users will not be able to influence this.
+    - This will ensure procedural macro compilation will be as hermetic as possible.
+- Scarb will use [`libloading`](https://crates.io/crates/libloading) for loading built procedural macros shared objects.
+
+### Procedural macro API in Cairo plugins
+
+- The procedural macro has to be called from Cairo code in order to be executed during the compilation. In contrast to current behaviour of Cairo plugins, no longer will they be executed on each node of AST.
+- The procedural macro will be triggered by one of three Cairo expressions
+    - Macro call, e.g. `xyz!`
+    - Macro attribute, e.g. `#[xyz]`
+    - Macro derive, e.g. `#[derive(Xyz)]`
+- The API for writing procedural macros for Cairo will be published for programmers, versioned separately from Scarb. This will provide a stability for the procedural macro implementation.
+- In total, the implementation will consist of three Rust crates, two of which will be visible by the macro author. These crates will be automatically added to the generated `Cargo.toml` file.
+  - First one, called `cairo-lang-macro-interface`, will contain the API definitions of the `TokenStream` and `ProcMacroResult` types used as input and output for macro implementation.
+  - The second one, called `cairo-lang-macro-attributes`, will contain implementation of Rust macros used for wrapping procedural macro entrypoint function. This will hide details of FFI communication from the procedural macro author.
```

Comment:
I think `-attributes` crate should be a dependency of the `-interface` one, and `-interface` should simply do `pub use _attributes::*`. That's what most crates do, for example `serde` vs `serde_derive`; or `clap`

Also, that way, the `-interface` part can be skipped imho

Reply:
Yep, reworded, thank you!